record-and-playback/core/lib/recordandplayback/edl/audio.rb

```diff
@@ -105,135 +105,132 @@ def self.render(edl, output_basename)
-              end
+        output
+      end
+
+      def self.process_segment(entry, output_basename, i, total_segments, audioinfo)
+        BigBlueButton.logger.info "Processing segment #{i} of #{total_segments} (duration: #{entry[:next_timestamp] - entry[:timestamp]} ms, num_audios: #{entry[:audios]&.length || 0})"
+        ffmpeg_cmd = [*FFMPEG]
+        filter_lines = []
+        track_labels = []
+        input_index = 0
+
+        if entry[:audios] && !entry[:audios].empty?
+          entry[:audios].each_with_index do |audio_data, idx|
+            filename = audio_data[:filename]
+            seek     = audio_data[:timestamp]
+            info     = audioinfo[filename]
+            speed    = 1.0
+
+            # Check for and handle audio files with mismatched lengths (generated
+            # by buggy versions of freeswitch in old BigBlueButton
+            if ((info[:format][:format_name] == 'wav' ||
+                 info[:audio][:codec_name] == 'vorbis') &&
+                 entry[:original_duration] &&
+                 (info[:duration].to_f / entry[:original_duration]) < 0.997 &&
+                 (((entry[:original_duration] - info[:duration]).to_f) / entry[:original_duration]).abs < 0.05)
+              speed = info[:duration].to_f / entry[:original_duration]
+              seek = 0
+              BigBlueButton.logger.warn "  Audio file length mismatch in #{filename}, adjusting speed to #{speed}"
             end
-      
-            # Mix audios if more than one
-            if track_labels.size > 1
-              # e.g.: [t0_0][t0_1]amix=inputs=2[seg0]
-              line = track_labels.join
-              line << "amix=inputs=#{track_labels.size}:normalize=0[#{seg_label}];"
+
+            # Skip this input and generate silence if the seekpoint is past the end of the audio, which can happen
+            # if events are slightly misaligned and you get unlucky with a start/stop or chapter break.
+            if seek < (info[:duration].to_f * speed)
+              # For each audio, add a -ss and -i for its input
+              ffmpeg_cmd += ['-ss', ms_to_s(seek)]
+              # Ensure that the entire contents of freeswitch wav files are read
+              if info[:format][:format_name] == 'wav'
+                ffmpeg_cmd += ['-ignore_length', '1']
+              end
+              # Prefer using the libopus decoder for opus files, it handles discontinuities better
+              if info[:audio][:codec_name] == 'opus'
+                ffmpeg_cmd += ['-c:a', 'libopus']
+              end
+              ffmpeg_cmd += ['-i', filename]
+
+              # Build track label
+              track_label = "t#{i}_#{idx}"
+              line = "[#{input_index}]#{FFMPEG_AFORMAT},apad,asetpts=N"
+              line << ",atempo=#{speed}" if speed != 1.0
+              line << "[#{track_label}];"
               filter_lines << line
+              track_labels << "[#{track_label}]"
+              input_index += 1
             else
```

Comment:
When looking through this, I've determined an issue that might cause audio desync.

If the audio file length mismatch check is triggered, above, it resets `seek = 0`. The reason for this is that the audio speed adjustment should be applied from the beginning of the file.

In order for this to work correctly, an atrim filter needs to be used *after the atempo filter* to remove audio prior to `audio_data[:timestamp]` - for example, use `",atempo=#{speed},atrim=start=#{ms_to_s(audio_data[:timestamp])}"`

Along with this change, the `asetpts=N` filter needs to be moved to after the `atrim` filter.

Reply:
Yes, that would work.

But i think it would be more clear to write something like
```ruby
line = "[#{input_index}]#{FFMPEG_AFORMAT},apad,asetpts=N"
line << ",atempo=#{speed},atrim=start=#{ms_to_s(audio_data[:timestamp])}" if speed != 1.0
line << ',asetpts=N'
```